who.py

```diff
@@ -8,279 +8,419 @@
+            logger.info(f"Populating DB for indicator {indicator_name}")
+            for row in indicator_json["value"]:
+                if row["SpatialDimType"] != "COUNTRY":
+                    continue
+                country_iso3 = row["SpatialDim"]
+                country_name = Country.get_country_name_from_iso3(country_iso3)
+                startyear = datetime.fromisoformat(
+                    row["TimeDimensionBegin"]
+                ).strftime("%Y")
+                endyear = datetime.fromisoformat(
+                    row["TimeDimensionEnd"]
+                ).strftime("%Y")
+                db_indicators_row = DBIndicatorData(
+                    id=row["Id"],
+                    indicator_code=indicator_code,
+                    indicator_name=indicator_name,
+                    indicator_url=indicator_url,
+                    year=row["TimeDim"],
+                    start_year=startyear,
+                    end_year=endyear,
+                    region_code=row["ParentLocationCode"],
+                    region_display=row["ParentLocation"],
+                    country_code=country_iso3,
+                    country_display=country_name,
+                    dimension_type=row["Dim1Type"],
+                    dimension_code=row["Dim1"],
+                    dimension_name=self._dimension_value_names_dict.get(
+                        row["Dim1"]
+                    ),
+                    numeric=row["NumericValue"],
+                    value=row["Value"],
+                    low=row["Low"],
+                    high=row["High"],
                 )
-                error = False
-                break
-            except RowError:
-                logger.warning("No PUBLISHSTATE (CODE)!")
-                error = True
-                sleep(600)
-        if error is True:
-            raise HDXError("WHO API has a problem!")
-        if success is True:
-            if len(all_rows) == 0:
-                headers = fileheaders
-                qcheaders = results["qcheaders"]
-                all_rows.extend(results["rows"])
-                qc_all_rows.extend(results["qcrows"])
-            else:
-                all_rows.extend(results["rows"][1:])
-                qc_all_rows.extend(results["qcrows"][1:])
-            for i, bite_disabled in enumerate(results["bites_disabled"]):
-                if bite_disabled is False:
-                    bites_disabled[i] = False
-    if len(all_rows) == 0:
-        logger.error(f"{countryname} has no data!")
-        return None, None, None
-
-    filename = f"health_indicators_{countryiso}.csv"
-    resourcedata = {
-        "name": f"All Health Indicators for {countryname}",
-        "description": "See resource descriptions below for links to indicator metadata",
-    }
-    dataset.generate_resource_from_rows(
-        folder, filename, all_rows, resourcedata, headers
-    )
+                self._session.add(db_indicators_row)
+            self._session.commit()
```

Comment:
Running the Python profiler showed that the most time is spent in commit followed by add (and then get_countryname_from_iso3 which seemed a bit odd to me as all it should be doing is a look up in a dictionary).

The commit and add time may be unavoidable, but one idea if a lot of rows are being committed all at once its to try changing to committing 1000 at a time with a log message which would give an idea of progress.

Reply:
I still need to profile it, but I'm giving [bulk insertion](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements) a go, I'll update once done